Index: linux-6.18.44/arch/arm64/boot/dts/rockchip/rk3588-base-pinctrl.dtsi
===================================================================
--- linux-6.18.44.orig/arch/arm64/boot/dts/rockchip/rk3588-base-pinctrl.dtsi
+++ linux-6.18.44/arch/arm64/boot/dts/rockchip/rk3588-base-pinctrl.dtsi
@@ -281,7 +281,9 @@
 		dp0m0_pins: dp0m0-pins {
 			rockchip,pins =
 				/* dp0_hpdin_m0 */
-				<4 RK_PB4 5 &pcfg_pull_none>;
+				<4 RK_PB4 5 &pcfg_pull_none>,
+				<2 RK_PB5 1 &pcfg_pull_none>,
+				<2 RK_PB6 1 &pcfg_pull_none>;
 		};
 
 		/omit-if-no-ref/
